### ISSUE TITLE
Issue during run

### ISSUE THREAD

**Issue by ntromas** [jiarong/VirSorter2#32]
Hi,

I got this error during VirSorter2 run (environmental metagenomes contigs):
virsorter run -i sequences_cdhit_metagenomes.fasta -w VIRSORTERvers2 -j 4 all -d /home/nico/miniconda3/envs/virsorter2/db/db
[2020-12-17 18:17 INFO] Using /home/nico/miniconda3/envs/virsorter2/lib/python3.8/site-packages/virsorter/template-config.yaml as config template
[2020-12-17 18:17 INFO] conig file written to /mnt/1bcce214-6fb0-454a-99a7-df65602d98b6/Metagenomic_2006_2018_reads/megahit_contigs/VIRSORTERvers2/config.yaml

[2020-12-17 18:17 INFO] Executing: snakemake --snakefile /home/nico/miniconda3/envs/virsorter2/lib/python3.8/site-packages/virsorter/Snakefile --directory /mnt/1bcce214-6fb0-454a-99a7-df65602d98b6/Metagenomic_2006_2018_reads/megahit_contigs/VIRSORTERvers2 --jobs 4 --configfile /mnt/1bcce214-6fb0-454a-99a7-df65602d98b6/Metagenomic_2006_2018_reads/megahit_contigs/VIRSORTERvers2/config.yaml --conda-prefix /home/nico/miniconda3/envs/virsorter2/db/db/conda_envs --rerun-incomplete --use-conda --nolock --latency-wait 600   --quiet  all   
Job counts:
	count	jobs
	1	all
	1	check_point_for_reclassify
	1	circular_linear_split
	1	classify
	5	classify_by_group
	5	classify_full_and_part_by_group
	1	combine_linear_circular
	5	combine_linear_circular_by_group
	1	extract_feature
	1	extract_provirus_seqs
	1	finalize
	1	gff_feature
	5	gff_feature_by_group
	5	hmm_features_by_group
	1	hmm_sort_to_best_hit_taxon
	5	hmm_sort_to_best_hit_taxon_by_group
	1	merge_classification
	1	merge_full_and_part_classification
	5	merge_hmm_gff_features_by_group
	5	merge_provirus_call_by_group_by_split
	1	merge_provirus_call_from_groups
	5	merge_split_hmmtbl
	25	merge_split_hmmtbl_by_group
	25	merge_split_hmmtbl_by_group_tmp
	1	pick_viral_fullseq
	1	preprocess
	1	split_faa
	5	split_faa_by_group
	5	split_gff_by_group
	121
[2020-12-17 18:18 INFO] # of seqs < 0 bp and removed: 0
[2020-12-17 18:18 INFO] # of circular seqs: 4226
[2020-12-17 18:18 INFO] # of linear seqs  : 2067163
[2020-12-17 18:18 INFO] Finish spliting circular contig file with common rbs
[2020-12-17 18:18 INFO] Finish spliting circular contig file with NCLDV rbs
[2020-12-17 18:18 INFO] Finish spliting linear contig file with common rbs
[2020-12-17 18:19 INFO] Finish spliting linear contig file with NCLDV rbs
[2020-12-17 20:01 INFO] Step 1 - preprocess finished.
/bin/bash: line 44: 2733402 Segmentation fault      (core dumped) hmmsearch -T 30 --tblout iter-0/NCLDV/all.pdg.faa.splitdir/all.pdg.faa.ss.120.split.Viruses.splithmmtbl --cpu 1 --noali -o /dev/null $Hmmdb $Tmp/$Bname 2> iter-0/NCLDV/all.pdg.faa.splitdir/all.pdg.faa.ss.120.split.Viruses.splithmm.log

[2020-12-20 04:46 ERROR] See error details in /mnt/1bcce214-6fb0-454a-99a7-df65602d98b6/Metagenomic_2006_2018_reads/megahit_contigs/VIRSORTERvers2/iter-0/NCLDV/all.pdg.faa.splitdir/all.pdg.faa.ss.120.split.Viruses.splithmm.log
[Sun Dec 20 04:46:05 2020]
Error in rule hmmsearch_by_group:
    jobid: 1351
    output: iter-0/NCLDV/all.pdg.faa.splitdir/all.pdg.faa.ss.120.split.Viruses.splithmmtbl
    log: iter-0/NCLDV/all.pdg.faa.splitdir/all.pdg.faa.ss.120.split.Viruses.splithmm.log (check log file(s) for error message)
    conda-env: /home/nico/miniconda3/envs/virsorter2/db/db/conda_envs/222d0847
    shell:
        
        Domain=Viruses
        Group_specific_hmmdb=/home/nico/miniconda3/envs/virsorter2/db/db/group/NCLDV/customized.hmm
        Rbs_pdg_db=/home/nico/miniconda3/envs/virsorter2/db/db/group/NCLDV/rbs-prodigal-train.db

        if [ $Domain = "Viruses" ]; then
            if [ -s $Group_specific_hmmdb ]; then
                Hmmdb=$Group_specific_hmmdb
            else
                Hmmdb=/home/nico/miniconda3/envs/virsorter2/db/db/hmm/viral/combined.hmm
            fi
        else
            Domain2=$Domain
            if [ $Domain2 = "Pfamviruses" ]; then
                Domain2=Viruses
            fi
            Hmmdb=/home/nico/miniconda3/envs/virsorter2/db/db/hmm/pfam/Pfam-A-"$Domain2".hmm
        fi

        if [ -s $Rbs_pdg_db ] || [ -s $Group_specific_hmmdb ]; then
            To_scratch=false
            Bname=$(basename iter-0/NCLDV/all.pdg.faa.splitdir/all.pdg.faa.ss.120.split)
            # move the heavy IO of hmmsearch in local scratch
            if [ -d "/tmp" ]; then
                # not sure df or du are compatible in all linux; use "||To_scratch=false" 
                #   to prevent imcompatibility in some linux distro
                Tmp=$(mktemp -d /tmp/vs2-XXXXXXXXXXXX) && To_scratch=true || To_scratch=false
                Avail=$(df -P /tmp | awk 'END{print $4}') || To_scratch=false
                Fsize=$(du -k iter-0/NCLDV/all.pdg.faa.splitdir/all.pdg.faa.ss.120.split | awk '{print $1*5}') || To_scratch=false
                if [ "$Avail" -gt "$Fsize" ] && [ "$To_scratch" = "true" ]; then
                    cp iter-0/NCLDV/all.pdg.faa.splitdir/all.pdg.faa.ss.120.split $Tmp/$Bname || To_scratch=false
                else
                    To_scratch=false
                fi
            fi

            if [ "$To_scratch" = false ]; then
                Inputseq=iter-0/NCLDV/all.pdg.faa.splitdir/all.pdg.faa.ss.120.split
            else
                # when To_scratch is true, Tmp and Bname should have been defined successfully
                hmmsearch -T 30 --tblout iter-0/NCLDV/all.pdg.faa.splitdir/all.pdg.faa.ss.120.split.Viruses.splithmmtbl --cpu 1 --noali -o /dev/null $Hmmdb $Tmp/$Bname 2> iter-0/NCLDV/all.pdg.faa.splitdir/all.pdg.faa.ss.120.split.Viruses.splithmm.log || { echo "See error details in /mnt/1bcce214-6fb0-454a-99a7-df65602d98b6/Metagenomic_2006_2018_reads/megahit_contigs/VIRSORTERvers2/iter-0/NCLDV/all.pdg.faa.splitdir/all.pdg.faa.ss.120.split.Viruses.splithmm.log" | python /home/nico/miniconda3/envs/virsorter2/lib/python3.8/site-packages/virsorter/./scripts/echo.py --level error; exit 1; }
                rm -f $Tmp/$Bname && rmdir $Tmp
            fi
        else
            touch iter-0/NCLDV/all.pdg.faa.splitdir/all.pdg.faa.ss.120.split.Viruses.splithmmtbl
        fi
        
        (one of the commands exited with non-zero exit code; note that snakemake uses bash strict mode!)

Exiting because a job execution failed. Look above for error message

Any idea? I used exactly the same input for VirSorter (1) and it works so I wonder if VirSorter2 required more memory?

Thanks for your help!

Cheers

Nico

**Comment by jiarong** [comment]
There has been similar error reports. It is likely to be caused by newer version of HMMER. Can you try the development version and see if the problem is fixed?

**Comment by ntromas** [comment]
Hi, 

It seems that I don't get that error anymore. However, Step 1 is still processing (Started the run the 23). 
Here is the details of the input file:
# number of contigs:     2071389
# total contigs length:  4964216520
# mean contig size:      2396.56
# contig size first quartile: 1177
# median contig size:         1490
# contig size third quartile: 2245
# longest contig:             695567
# shortest contig:            1000

VirSorter1 finished the job in ~4 days. Is it normal that this second version takes much more time?

Thanks for your help,

**Comment by jiarong** [comment]
Yes, VirSorter2 needs much more computation time due to larger annotation databases.  More threads (`-j`) can be added to speed up the run.

**Comment by ntromas** [comment]
Hi,

Ok it worked perfectly now. I have another naive question, is VirSorter2 searching for prophage too? It is not super clear for me from the output.

Cheers,

**Comment by jiarong** [comment]
Sequence names ending with "_partial" are prophages.

**Comment by ntromas** [comment]
Thanks!
I would have a last question, to avoid confusion, how the max score is calculated. Cause I can get max score=1 and viral% <40. Just want to understand properly the output (I am using environmental metagenomes) to define a good cut-off for score.
Thanks a lot for your help!

**Comment by jiarong** [comment]
The score is predicted using a model trained from known (RefSeq) viral and cellular sequences. That means some known viral sequences have max_score=1 and viral% < 40. Also note that many viruses have large % of genes that can not annotated.
